include/soci/soci-backend.h

```diff
@@ -319,6 +322,68 @@ class blob_backend
     SOCI_NOT_COPYABLE(blob_backend)
 };
 
+/**
+ * This Blob implementation uses an explicit buffer that is read from and written to, instead of
+ * directly communicating with the underlaying database.
+ * Thus, it is intended to be used whenever the underlying database does not offer a more efficient
+ * way of dealing with BLOBs.
+ */
+class trivial_blob_backend : public blob_backend
+{
+public:
+    std::size_t get_len() override { return buffer_.size(); }
+
+    std::size_t read_from_start(char* buf, std::size_t toRead,
+        std::size_t offset = 0) override
+    {
+        if (offset > buffer_.size() || (offset ==  buffer_.size() && offset > 0))
```

Comment:
Unless this is necessary for compatibility with the existing behaviour, I'd _really_ like to change this. What is the rationale for allowing successfully reading from an empty blob? None of the data containers I know of behave like this. You definitely can't read from an empty file, or stream, which is what blob object emulates.

Reply:
It's the same rational of allowing a specified size bigger than the actual size of the blob. And that is indeed required for backwards compatibility.

> None of the data containers I know of behave like this. You definitely can't read from an empty file, or stream, which is what blob object emulates.

These also error when I try to read even a single byte more than they contain. `blob` does not follow that either

Reply:
>  whether it's normal for a BLOB to be empty

why not? to me this seems equally normal as encountering an empty `vector` or an empty file. Sure, these things are intended to store data but they might just as well contain no data. More specific to the blob example could be a table that uses a NOT NULL constraint on the blob column, but still wishes to represent empty datasets.

> If this is not the case, I'd like to understand where do these empty blobs come from.

Keeping above database example in mind, such an empty blob can be the result of querying from the DB. Afaik there is nothing that prevents folks to store empty blobs in a database.

> No, this is wrong. You can (try to) read more data from a file than it contains without problems, semantics of standard read() is well understood and there is nothing wrong with providing it with a larger buffer than necessary.

Exactly. I can also read an empty file without having an exception thrown at me.
My quote from above (about not being able to read more than is there) was referring to things like `std::vector`. And because of what you wrote, I would argue that a blob is semantically closer to a file than something like a vector and should therefore support reading operations on an empty object.

Reply:
For me reading from an empty blob is like reading from a file already at EOF and IMO it should fail. But if you feel so strongly about it, let's keep the current code, just please make sure to mention that reading from empty blob doesn't fail because, again, I think this is very unexpected. TIA!

Reply:
> just please make sure to mention that reading from empty blob doesn't fail because, again, I think this is very unexpected. TIA!

where do you want me to mention this? In the LOB docs?

Reply:
Wherever you think it's best, I'd just like to be able to point people who ask about this in the future to the part of the docs where this is clearly documented.

But, yes, LOB docs looks like a good place.

TIA!